### PR TITLE
feat: add case number validation to 'send to gatekeeper'

### PR DESCRIPTION
**Pull request by joez-f**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-1000

### Change description ###

Add validation to the send to gatekeeper event to ensure that a case number has been added. This is crucial to enable C6 in gatekeeping state as C6 requires a case number, and a case number cannot be added during gatekeeper so we will validate that a case has a case number before it can progress to gatekeeping. e2e tests updated

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java [hunk 1]
```diff
@@ -49,16 +53,26 @@
     @Autowired
     private ObjectMapper mapper;
 
+    @Test
+    void shouldReturnErrorsWhenFamilymanNumberIsNotProvided() throws Exception {
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: Can we caps 'FamilyMan' to keep consistent?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java [hunk 2]
```diff
@@ -49,16 +53,26 @@
...
+        CallbackRequest request = CallbackRequest.builder()
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: Could we extract to private method?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java [hunk 3]
```diff
@@ -49,16 +53,26 @@
...
+            .caseDetails(CaseDetails.builder()
+                .id(12345L)
+                .data(ImmutableMap.of("data", "some data"))
+                .build())
+            .build();
+
+        MvcResult mvc = makeRequest(request, "about-to-start");
+
+        AboutToStartOrSubmitCallbackResponse callbackResponse = mapper.readValue(mvc.getResponse()
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: Could extract this to private method also

File: e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js [hunk 4]
```diff
@@ -20,9 +20,12 @@ Before(async (I, caseViewPage, submitApplicationEventPage, sendCaseToGatekeeperE
 
     I.signOut();
 
-    //hmcts login and send to gatekeeper
+    //hmcts login, enter case number and send to gatekeeper
     await I.signIn(config.hmctsAdminEmail, config.hmctsAdminPassword);
     await I.navigateToCaseDetails(caseId);
+    await caseViewPage.goToNewActions(config.administrationActions.addFamilyManCaseNumber);
```
Comment (toireasam) [inline e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js]: This shouldn't need await infront of it

File: e2e/tests/gatekeeperMakesAllocationDecision_test.js [hunk 5]
```diff
@@ -19,9 +19,12 @@ Before(async (I, caseViewPage, submitApplicationEventPage, sendCaseToGatekeeperE
 
     I.signOut();
 
-    //hmcts login and send to gatekeeper
+    //hmcts login, enter case number and send to gatekeeper
     await I.signIn(config.hmctsAdminEmail, config.hmctsAdminPassword);
     await I.navigateToCaseDetails(caseId);
+    await caseViewPage.goToNewActions(config.administrationActions.addFamilyManCaseNumber);
```
Comment (toireasam) [inline e2e/tests/gatekeeperMakesAllocationDecision_test.js]: This shouldn't need await infront of it

File: e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js [hunk 6]
```diff
@@ -20,9 +20,12 @@ Before(async (I, caseViewPage, submitApplicationEventPage, sendCaseToGatekeeperE
 
     I.signOut();
 
-    //hmcts login and send to gatekeeper
+    //hmcts login, add case number and send to gatekeeper
     await I.signIn(config.hmctsAdminEmail, config.hmctsAdminPassword);
     await I.navigateToCaseDetails(caseId);
+    await caseViewPage.goToNewActions(config.administrationActions.addFamilyManCaseNumber);
```
Comment (toireasam) [inline e2e/tests/localAuthorityCompliesWithDirectionsAfterSDOIsIssued_test.js]: This shouldn't need await infront of it

File: e2e/tests/localAuthorityMaintainsCaseAfterGatekeeping_test.js [hunk 7]
```diff
@@ -19,9 +19,12 @@ Before(async (I, caseViewPage, submitApplicationEventPage, sendCaseToGatekeeperE
 
     I.signOut();
 
-    //hmcts login and send to gatekeeper
+    //hmcts login, enter case number and send to gatekeeper
     await I.signIn(config.hmctsAdminEmail, config.hmctsAdminPassword);
     await I.navigateToCaseDetails(caseId);
+    await caseViewPage.goToNewActions(config.administrationActions.addFamilyManCaseNumber);
```
Comment (toireasam) [inline e2e/tests/localAuthorityMaintainsCaseAfterGatekeeping_test.js]: This shouldn't need await infront of it

File: e2e/tests/localAuthorityMaintainsCaseAfterSDO_test.js [hunk 8]
```diff
@@ -21,9 +21,12 @@ Before(async (I, caseViewPage, submitApplicationEventPage, sendCaseToGatekeeperE
 
     I.signOut();
 
-    //hmcts login and send to gatekeeper
+    //hmcts login, add case number and send to gatekeeper
     await I.signIn(config.hmctsAdminEmail, config.hmctsAdminPassword);
     await I.navigateToCaseDetails(caseId);
+    await caseViewPage.goToNewActions(config.administrationActions.addFamilyManCaseNumber);
```
Comment (toireasam) [inline e2e/tests/localAuthorityMaintainsCaseAfterSDO_test.js]: This shouldn't need await infront of it

**Comment by toireasam** [comment]
Couple of minor comments, otherwise looks good. :)

File: e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js (hunk 4 above)
Comment (joez-f) [reply e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js]: Will remove and rerun e2e tests

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java (hunk 1 above)
Comment (joez-f) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: C6 and C21 use 'Familyman' so we'll leave it like this for now to be consistent with them, as I don't want to change those files in this PR

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java [hunk 9]
```diff
@@ -79,15 +83,33 @@ void shouldBuildGatekeeperNotificationTemplateWithCompleteValues() throws Except
...
-            .andExpect(status().isOk());
+                .content(mapper.writeValueAsString(request)))
+            .andExpect(status().isOk())
+            .andReturn();
+    }
 
-        verify(notificationClient, times(1)).sendEmail(GATEKEEPER_SUBMISSION_TEMPLATE, GATEKEEPER_EMAIL,
-            expectedGatekeeperParameters, "12345");
+    private AboutToStartOrSubmitCallbackResponse requestWithoutCaseNumber() throws Exception {
+        CallbackRequest request = CallbackRequest.builder()
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: I think i'd keep this method as 'buildCallbackRequest' that returns a CallbackRequest and move the rest of it out, lets discuss on slack

**Comment by toireasam** [comment]
LGTM

**Pull request merged**